analyzer/tools/build-logger/src/ldlogger-util.c

```diff
@@ -112,30 +115,84 @@ char* shellEscapeStr(const char* str_, char* buff_)
   {
     switch (*str_)
     {
+      case '\a':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 'a';
+        break;
+      case 0x1B: // \e
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 'e';
+        break;
       case '\t':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 't';
+        break;
       case '\b':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 'b';
+        break;
       case '\f':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 'f';
+        break;
       case '\r':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 'r';
+        break;
       case '\v':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 'v';
+        break;
       case '\n':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = 'n';
+        break;
       case ' ':
         *out++ = '\\';
         *out++ = '\\';
-        *out++ = *str_++;
+        *out++ = ' ';
         break;
-
       case '\\':
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = '\\';
+        *out++ = '\\';
+        break;
       case '\"':
         *out++ = '\\';
         *out++ = '\\';
         *out++ = '\\';
-        *out++ = *str_++;
+        *out++ = '\"';
         break;
-
-      default:
-        *out++ = *str_++;
+      default: {
+        // Escape the rest of the control characters.
+        unsigned char value = *str_;
+        if (value < 0x20) {
+          static const unsigned char dec2hex_last_digit[32] = {
+            '0','1','2','3','4','5','6','7','8','9', 'A', 'B', 'C', 'D', 'E', 'F',
+            '0','1','2','3','4','5','6','7','8','9', 'A', 'B', 'C', 'D', 'E', 'F',
+          };
+          *out++ = '\\';
+          *out++ = '\\';
+          *out++ = 'x';
+          *out++ = (value < 16 ? '0' : '1');
+          *out++ = dec2hex_last_digit[value];
```

Comment:
The value is not checked here and it can be greater than the length of the `dec2hex_last_digit` array.

Reply:
No, it's safe. The enclosing if condition guards this. (Note that `value` is unsigned, so the condition actually checks the right thing even if the character was negative originally.